03-client-api/04-new.md

```diff
@@ -0,0 +1,287 @@
+* Can be compiled to a variety of language stubs.
+
+Our clients don't use any complex components of gRPC like channel multiplexing. Instead, we take full advantage
+of the ability to create complex and strongly typed messages that are streamed between the client and the server.
+
+### Understanding the Grakn gRPC protocol
+
+The full protocol definition is in our [git repository](https://github.com/graknlabs/grakn/tree/master/protocol).
+
+To become familiar with the protocol structure, let's look at the following excerpt from `Session.proto`. We will be looking 
+at the process of getting attributes of a specific value from a transaction - the relevant RPC messages are in **bold**.
+
+```
+service SessionService {
+    ...
+    rpc transaction (stream Transaction.Req) returns (stream Transaction.Res);
+}
+...
+
+message Transaction {
+    message Req {
+        map<string, string> metadata = 1000;
+        oneof req {
+            Open.Req open_req = 1;
+            Commit.Req commit_req = 2;
+            Query.Req query_req = 3;
+            **Iter.Req iterate_req = 4;**
+            GetSchemaConcept.Req getSchemaConcept_req = 5;
+            GetConcept.Req getConcept_req = 6;
+            **GetAttributes.Req getAttributes_req = 7;**
+            ...
+         }
+    }
+    message Res {
+            oneof res {
+                Open.Res open_res = 1;
+                Commit.Res commit_res = 2;
+                Query.Iter query_iter = 3;
+                **Iter.Res iterate_res = 4;**
+                GetSchemaConcept.Res getSchemaConcept_res = 5;
+                GetConcept.Res getConcept_res = 6;
+                **GetAttributes.Iter getAttributes_iter = 7;**
+                   ...
+            }
+    }
+    ...
+    message Iter {
+        message Req {
+            int32 id = 1;
+        }
+        message Res {
+            oneof res {
+                bool done = 1;
+                Query.Iter.Res query_iter_res = 2;
+                **GetAttributes.Iter.Res getAttributes_iter_res = 3;**
+                Method.Iter.Res conceptMethod_iter_res = 4;
+            }
+        }
+    }
+    ...
+    message GetAttributes {
+        message Req {
+            ValueObject value = 1;
+        }
+        message Iter {
+            int32 id = 1;
+            message Res {
+                Concept attribute = 1;
+            }
+        }
+    }
+```
+
+This definition tells us that we have one `transaction` RPC endpoint, which is a bi-directional streams that takes messages
+of type `Transaction.Req` and returns messages of type `Transaction.Res`. These types permit a variety of different 
+bodies, and we can trace through the definition to determine how to build messages, and what the expected responses are.
+To understand what exactly this means, let's look at a detailed example.
+
+
+### Get Attributes by Value 
+Since all our clients are implemented similarly, the following piece of Python is representative:
+```
```

Comment:
this code block can be made testable.

we just need to make sure that it's using the `social_network` as the target keyspace, and then add 
`<!-- test-example <an-appropriate-unique-file-title>.py -->` on the line before the opening ` ```python`